Update TypeScript type definitions for server options

joelmukuthu opened the pull request:
- Adds back definitions that were inadvertently removed in https://github.com/fastify/fastify/pull/1569
- Adds tests for server-option definitions
- Updates server-option definitions to match code and docs in master
- Fixes an error message for AJV plugin options 
- Adds a new `FastifyLoggerInstance` type 
- Removes `FastifyLoggerInstance` methods from `FastifyLoggerOptions`

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Tip: `npm run bench` to compare branches interactively.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md
-->

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

Ethan-Arrowood commented on fastify.d.ts:
```diff
@@ -71,28 +74,45 @@ export type FastifyServerOptions<
   Logger = FastifyLoggerOptions<RawServer>
 > = {
+  http2?: boolean,
+  https?: https.ServerOptions
```
These were specifically removed for the discriminant union overloading. Is there an issue with the types as is?

joelmukuthu replied:
Right, good catch. Fixing

Ethan-Arrowood commented on fastify.d.ts:
```diff
@@ -71,28 +74,45 @@ export type FastifyServerOptions<
+    customOptions?: ajv.Options,
+    plugins?: Function[]
+  },
+  http2SessionTimeout?: number,
```
this should be added to the http2 options interfaces

Ethan-Arrowood commented on test/types/fastify.test-d.ts:
```diff
@@ -18,3 +18,51 @@ expectType<LightMyRequestChain>(fastify({ http2: true, https: {}}).inject())
 expectError(fastify<http2.Http2SecureServer>({ http2: false })) // http2 option must be true
+
+// server options
+fastify({ http2: true })
```
Can these be wrapped with expectType or expectAssignable?

joelmukuthu replied:
Do you mean `expectType<FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse>>(fastify({ http2: true }))`? In the spirit of unit-testing I left that out as it's tested [elsewhere](https://github.com/fastify/fastify/pull/2277/files/5867d26d5faa08fa0839f62f14ee13c6ebb58b34#diff-7bd3eec77e6d6f08f2506c0170ce032eR10)

Ethan-Arrowood replied:
Ahh okay I think you should simplify these tests into one:

```
expectAssignable<FastifyInstance>(fastify({
  // ... add as many options as possible here
})
```

joelmukuthu replied:
I split them up to guard against options that are accidentally defined as required:

```ts
interface ServerOptions {
  ignoreTrailingSlash?: boolean,
  connectionTimeout: number, // bad, should be `connectionTimeout?: number`
} 
```

In this case, a single test wouldn't catch that:

```ts
fastisfy({
  ignoreTrailingSlash: true,
  connectionTimeout: 1000
})
```

... and also unit-testing :)

Ethan-Arrowood replied:
Okay then in this case I'd like to see 

```ts
expectAssignable<FastifyInstance>(fastify({ option: value })
```

for every option as this asserts the given instance is returning what we expect

joelmukuthu replied:
Cool. That works for every case except when any `logger` option is set (e.g. `expectAssignable<FastifyInstance>(fastify({ logger: true }))`), where it fails with: 

```text
Argument of type 'FastifyInstance<Server, IncomingMessage, ServerResponse, true>' is not assignable to parameter of type 'FastifyInstance<Server, IncomingMessage, ServerResponse, FastifyLoggerOptions<Server, IncomingMessage, ServerResponse>>'.
  Type 'true' is not assignable to type 'FastifyLoggerOptions<Server, IncomingMessage, ServerResponse>'
```

How can I fix that?

Ethan-Arrowood replied:
Because the logger type is inferred by the generics, you'll need to use `expectType<FastifyInstance<http.server, http.incomingmessage, http.serverresponse, true>` to test this option

joelmukuthu replied:
Thanks for the help, fixed.

joelmukuthu commented:
@mcollina @Ethan-Arrowood apologies for adding new changes after your reviews, I just stumbled upon the missing `logger` options. This should be ready for review now.

Ethan-Arrowood commented on package.json:
```diff
@@ -147,6 +147,7 @@
     "yup": "^0.28.4"
   },
   "dependencies": {
+    "@types/pino": "^6.0.1",
```
we intentionally omitted these types and provide instructions how to include them in the typescript documentation (see example 5 here: https://www.fastify.io/docs/latest/TypeScript/#fastify).

mcollina replied:
I totally missed this. Definitely -1. Those types are not really maintained.

joelmukuthu replied:
Thanks for the info. The TypeScript docs are incredibly well written and detailed, but my suspicion is that this might come up again as users might expect some level of completion because Fastify ships with its own types. Anyway I removed the pino types and added a comment in code referencing the example.

Ethan-Arrowood commented on types/logger.d.ts:
```diff
@@ -1,5 +1,7 @@
 import { FastifyError } from './error'
 import { RawServerBase, RawServerDefault, RawRequestDefaultExpression, RawReplyDefaultExpression } from './utils'
+import { FastifyRequest } from './request'
+import { LoggerOptions as PinoLoggerOptions } from 'pino'
```
please see previous comment about intentionally omitting pino types

joelmukuthu replied:
Thanks for the heads up! I've removed the pino types.

Ethan-Arrowood commented on types/logger.d.ts:
```diff
@@ -11,14 +13,24 @@ export interface FastifyLogFn {
+
 /**
  * Fastify Custom Logger options.
  */
```
At one point there was a comment here that explained what do for Pino types. Maybe we should add the back in if this is misleading.

joelmukuthu replied:
I added a new comment :)

joelmukuthu commented:
There's an issue with the current type definition for `Fastify.prototype.log` that I don't know how to fix, @Ethan-Arrowood let me know if you have some pointers. You can see the failing test here https://github.com/fastify/fastify/pull/2277/checks?check_run_id=673774170#step:6:9387

Ethan-Arrowood commented on test/types/fastify.test-d.ts:
```diff
@@ -12,9 +13,71 @@ expectType<FastifyInstance<http.Server, http.IncomingMessage, http.ServerRespons
+expectAssignable<FastifyInstance>(fastify({ onProtoPoisoing: 'error' }))
+expectAssignable<FastifyInstance>(fastify({ onConstructorPoisoning: 'error' }))
+expectAssignable<FastifyInstance>(fastify({ logger: true }))
+expectAssignable<FastifyInstance>(fastify({
```
for what its worth, I believe we are already testing the logger option in the logger type def test file. these are redundant here. Nice to have for sure but not necessary.

In order to get them to pass you need to pass the same logger type to the fastify instance generic paramter (thus also passing the http generics). Ik this is not ideal but its whats necessary

joelmukuthu replied:
Thanks for the help. I decided to keep the tests as they provide examples of how to overload logger options with various types and they also helped fix some bugs in the definitions for `FastisfyLoggerInstance['child']` and `FastifyLoggerOptions`.

joelmukuthu commented:
Updated and tests fixed, PTAL @Ethan-Arrowood @mcollina. I also updated the PR description [up top](#issue-417438119) to detail out the changes to logger options.
